src/tasks/tasks.hpp

```diff
@@ -0,0 +1,493 @@
+  std::unordered_set<Task *> dependencies;
+  std::pair<int, int> exec_limits;
+  TaskType task_type = TaskType::normal;
+  int num_calls = 0;
+  TaskStatus task_status = TaskStatus::incomplete;
+  std::mutex mutex;
+};
+
+class TaskRegion;
+class TaskList {
+  friend class TaskRegion;
+
+ public:
+  TaskList() : TaskList(TaskID(), {1, 1}) {}
+  explicit TaskList(const TaskID &dep, std::pair<int, int> limits)
+      : dependency(dep), exec_limits(limits) {
+    // make a trivial first_task after which others will get launched
+    // simplifies logic for iteration and startup
+    tasks.push_back(std::make_shared<Task>(
+        dependency,
+        [&tasks = tasks]() {
+          for (auto &t : tasks) {
+            t->SetStatus(TaskStatus::incomplete);
+          }
+          return TaskStatus::complete;
+        },
+        exec_limits));
+    first_task = tasks.back().get();
+    // connect list dependencies to this list's first_task
+    for (auto t : first_task->GetDependencies()) {
+      t->AddDependent(first_task, TaskStatus::complete);
+    }
+
+    // make a trivial last_task that tasks dependent on this list's execution
+    // can depend on.  Also simplifies exiting completed iterations
+    tasks.push_back(std::make_shared<Task>(
+        TaskID(),
+        [&completion_tasks = completion_tasks]() {
+          for (auto t : completion_tasks) {
+            t->reset_iteration();
+          }
+          return TaskStatus::complete;
+        },
+        exec_limits));
+    last_task = tasks.back().get();
+  }
+
+  template <class... Args>
+  TaskID AddTask(TaskID dep, Args &&...args) {
+    return AddTask(TaskQualifier::normal, dep, std::forward<Args>(args)...);
+  }
+
+  template <class... Args>
+  TaskID AddTask(const TaskQualifier tq, TaskID dep, Args &&...args) {
+    assert(tq.Valid());
+
+    // user-space tasks always depend on something. if no dependencies are given,
+    // make the task dependent on the list's first_task
+    if (dep.empty()) dep = TaskID(first_task);
+
+    if (!tq.Once() || (tq.Once() && unique_id == 0)) {
+      AddUserTask(dep, std::forward<Args>(args)...);
+    } else {
+      tasks.push_back(std::make_shared<Task>(
+          dep, [=]() { return TaskStatus::complete; }, exec_limits));
+    }
+
+    Task *my_task = tasks.back().get();
+    TaskID id(my_task);
+
+    if (tq.LocalSync() || tq.GlobalSync() || tq.Once()) {
+      regional_tasks.push_back(my_task);
+    }
+
+    if (tq.GlobalSync()) {
+      bool do_mpi = false;
+#ifdef MPI_PARALLEL
+      // make status, request, and comm for this global task
+      global_status.push_back(std::make_shared<int>(0));
+      global_request.push_back(std::make_shared<MPI_Request>(MPI_REQUEST_NULL));
+      // be careful about the custom deleter so it doesn't call
+      // an MPI function after Finalize
+      global_comm.emplace_back(new MPI_Comm, [&](MPI_Comm *d) {
+        int finalized;
+        MPI_Finalized(&finalized);
```

Comment:
I think the MPI calls should be wrapped in `PARTHENON_MPI_CHECK`

Reply:
Yep.  I'll do that.